addons/hr_contract/models/hr_contract.py

```diff
@@ -172,17 +180,31 @@ def update_state(self):
                 ('date_start', '>', contract.date_start)
             ], order="date_start asc", limit=1)
             if next_contract:
-                contract.date_end = next_contract.date_start - relativedelta(days=1)
+                contract.safe_write_for_cron({'date_end': next_contract.date_start - relativedelta(days=1)}, from_cron)
                 continue
             next_contract = self.search([
                 ('employee_id', '=', contract.employee_id.id),
                 ('date_start', '>', contract.date_start)
             ], order="date_start asc", limit=1)
             if next_contract:
-                contract.date_end = next_contract.date_start - relativedelta(days=1)
+                contract.safe_write_for_cron({'date_end': next_contract.date_start - relativedelta(days=1)}, from_cron)
 
         return True
 
+    def safe_write_for_cron(self, vals, from_cron=False):
```

Comment:
Here I trying something like 
```suggestion
    def safe_write_for_cron(self, vals, from_cron=False):
        if from_cron:
            try:
                with self.env.cr.savepoint():
                    self.write(vals)
            except ValidationError:
                self.env.cr.rollback()
                for contract in self: # if we do this, here we have cacheMiss.. That contract record does not exsist
                    try:
                        with self.env.cr.savepoint():
                            contract.write(vals)
                    except ValidationError as e:
                        _logger.warning(e)
                        self.env.cr.rollback()
                    else:
                        self.env.cr.commit()
            else:
                self.env.cr.commit()
        else:
            self.write(vals)
```

Seems like it would have been better for performance, but this does not do a trick